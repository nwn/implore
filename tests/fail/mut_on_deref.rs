use implore::impl_op;

#[derive(Copy, Clone, PartialEq, Eq, Debug)]
struct X;

#[impl_op]
fn deref(x: &X) -> X {
    X
}

#[impl_op]
fn deref(x: X) -> &X {
    &mut X
}

#[impl_op]
fn deref(x: &mut X) -> &X {
    &X
}

#[impl_op]
fn deref(x: &X) -> &mut X {
    &mut X
}

fn main() {}
