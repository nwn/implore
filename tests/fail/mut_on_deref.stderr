error: `deref` must return a reference type
 --> $DIR/mut_on_deref.rs:7:20
  |
7 | fn deref(x: &X) -> X {
  |                    ^

error: `deref` must take a reference
  --> $DIR/mut_on_deref.rs:12:13
   |
12 | fn deref(x: X) -> &X {
   |             ^

error: `deref` must take an immutable reference
  --> $DIR/mut_on_deref.rs:17:13
   |
17 | fn deref(x: &mut X) -> &X {
   |             ^^^^^^

error: `deref` must return an immutable reference type
  --> $DIR/mut_on_deref.rs:22:20
   |
22 | fn deref(x: &X) -> &mut X {
   |                    ^^^^^^
